### PR TITLE
Add fwupd support

### PR DESCRIPTION
**Pull request by sharadhr**
A function, `sign_fwupd` is added.

This function firstly checks that `fwupd` is installed and the service running. If so, the function checks if `/usr/lib/fwupd/efi/fwupdx64.efi.signed` exists, and exits if true. Otherwise, the signed file is generated with the user’s keys, and `fwupd.service` is restarted.

### CODE REVIEW

File: sbupdate [hunk 1]
```diff
@@ -134,6 +134,30 @@ function remove_image() {
...
+    echo "fwupd package is installed, and service is running."
+    if sbverify --cert "${KEY_DIR}"/@(DB|db).crt "${fwupd_dir}/${img_name}".signed > /dev/null; then
+      echo "Skipping already signed file ${fwupd_dir}/${img_name}."
+    else
+      # Sign the fwupd image file.
+      echo "Signing ${fwupd_dir}/${img_name}."
+      sbsign --key "${KEY_DIR}"/@(DB|db).key --cert "${KEY_DIR}"/@(DB|db).crt "${fwupd_dir}/${img_name}"
+      # Restart the fwupd service.
+      echo "Restarting fwupd.service."
+      systemctl restart fwupd.service --quiet > /dev/null
```
Comment (fenik) [inline sbupdate]: I have a question regarding this line: Is it necessary to restart the service, when the image has been signed?

**Comment by fenik** [comment]
One case that isn't covered by the changes are updates to the fwupdx64.efi through pacman. At the moment I've this (barebones) pacman hook running locally, that I think is worth sharing:

```
[Trigger]
Operation = Install
Operation = Upgrade
Type = File
Target = usr/lib/fwupd/efi/fwupdx64.efi

[Action]
Description = Signing fwupd EFI loader...
When = PostTransaction
Exec = /usr/bin/sbsign --key /etc/efi-keys/db.key --cert /etc/efi-keys/db.crt /usr/lib/fwupd/efi/fwupdx64.efi
Depends = sbsigntools
```

Adding something along those lines should sign new versions of fwupdx64.efi whenever they are added through pacman updates.

File: sbupdate (hunk 1 above)
Comment (sharadhr) [reply sbupdate]: When I tried running `fwupdmgr get-devices` after signing the image, there was no change in the behaviour, and my notebook's firmware was still marked as unsigned. Restarting the service solved the problem, so I implemented this behaviour into the script.

**Comment by rustonaut** [comment]
Hy,

I'm wondering if it makes sense to add the `[Trigger]` mentioned above to the `hooks/95-sbupdate.hook` file. Currently (in this PR) every time sbupdate runs the fwupd binary is potentially signed, so why not just run it every time that binary is updated by default. And if fwudp is not installed the trigger will just never trigger. Sure this might cause unnecessary signing (e.g. linux update but no fwupd update or the other way around). But I don't think that this really matters in practice and sometimes having a simpler solution is better. (Alternatively we can run the linux boot sign part and the fwudp sign part seperately based on subcommands, but then we maybe might want to also run fwudp implicitly on the initial run of sbupdate).

Also do you know if `/usr/lib/fwupd/efi/fwupdx64.efi.signed` is automatically removed when `/usr/lib/fwupd/efi/fwupdx64.efi` is update. Else the "skip check" would need to be updated to also not skip if e.g. the write/creation timestamp of `fwupdx64.efi` is newer then `fwupdx64.efi.signed`.

Besides that thanks for the work you did on this.

**Comment by rustonaut** [comment]
Another think I was wondering if we really only should sign it if the systemd service is active. 

The signing is about weather or not fwudp _can run_ but is-active is about if it _does/want's to run_, assuming it can run. 

I personally would sign the binary independent of weather or not the systemd service is active (but only restart it if it's active). I think that if we only sign if the service is active people might run into problems when they temporary have the service disabled update their system then they re-enable the service and it doesn't work because re-enabling the service doesn't cause the sign hook to run (nor should it cause it IMHO).

**Comment by fenik** [comment]
Hey @rustonaut!

I've just checked: Removing fwupd does not remove fwupdx64.efi.signed. However as the filename and path for it are hardcoded (or set per default) in fwupd, I think that the removal of the file should be part of the Arch fwupd package.

As for the other topic: In my opinion the EFI should only be signed whenever it's changed. Adding a separate pacman hook and sub-routine in sbupdate should not be a hurdle, but possibly against the current design of the sbupdate script.

Additionally I would not care about whether or not the fwupd.service is active when doing the signing, but I agree that we should only restart it, if it is active.

Anyways, I'd like @andreyv to chime in and help us out here with some kind of direction. I really want to see this feature in sbupdate and I'll gladly help wherever and however I can.

Cheers and thanks to all

**Comment by andreyv** [comment]
@fenik I completely agree with your assessment, and this is what I now implemented in b51ba62f6522d0c2c70ddd3b5e2254a245e58bf0.

Reloading a service in the hook would be done with
```shell
/usr/share/libalpm/scripts/systemd-hook reload fwupd
```
It won't do anything if the service is not active. However, from fwupd source code it seems that it should pick up the updated file automatically.

**Comment by andreyv** [comment]
Closing as per https://github.com/andreyv/sbupdate/pull/25#issuecomment-714266490.

**Pull request closed without merging**